subprojects/testing-base/src/main/java/org/gradle/api/tasks/testing/AbstractTestTask.java

```diff
@@ -481,11 +485,25 @@ public void execute(DeleteSpec spec) {
     private void handleCollectedResults(TestCountLogger testCountLogger) {
         if (testCountLogger.hadFailures()) {
             handleTestFailures();
-        } else if (testCountLogger.getTotalTests() == 0 && shouldFailOnNoMatchingTests()) {
-            throw new TestExecutionException(createNoMatchingTestErrorMessage());
+        } else if (testCountLogger.getTotalTests() == 0) {
+            if (shouldFailOnNoMatchingTests()) {
+                throw new TestExecutionException(createNoMatchingTestErrorMessage());
+            } else if (!successWithoutTest.isPresent()) {
+                emitDeprecationMessage();
+            } else if (!successWithoutTest.get()) {
+                throw new TestExecutionException(createNoMatchingTestErrorMessage());
+            }
```

Comment:
❌ For Predictive Test Selection, we'll need a way to avoid failing here if there were > 0 tests discovered but 0 were selected.

Reply:
Thanks Marc
How could we achieve this in the best way for you?

Reply:
I think the `TestExecuter` would need a way to signal the number of tests it discovered, probably via a new method on `TestResultProcessor`. I've already reached out to @FrauBoes internally and we agreed to discuss options when she picks up work on this PR again.

Reply:
@marcphilipp Please see #24786 for work on the test task option. Note that I propose not to deprecate the current behaviour so it should not affect Predictive Test Selection anymore.